packages/blockchain/src/blockchain.ts

```diff
@@ -369,89 +371,177 @@ export class Blockchain implements BlockchainInterface {
-        const currentTd = { header: BIGINT_0, block: BIGINT_0 }
-        let dbOps: DBOp[] = []
-
         if (block.common.chainId() !== this.common.chainId()) {
           throw new Error(
             `Chain mismatch while trying to put block or header. Chain ID of block: ${block.common.chainId}, chain ID of blockchain : ${this.common.chainId}`,
           )
         }
 
-        if (this._validateBlocks && !isGenesis && item instanceof Block) {
-          // this calls into `getBlock`, which is why we cannot lock yet
-          await this.validateBlock(block)
-        }
+        const { header } = block
+        const blockHash = header.hash()
+        const blockNumber = header.number
 
-        if (this._validateConsensus) {
-          await this.consensus!.validateConsensus(block)
-        }
+        // check if head is still canonical i.e. if this is a block insertion on tail or reinsertion
+        // on the same canonical chain
+        // let isHeadChainStillCanonical
+        // if (opts?.canonical !== false) {
+        //   const childHeaderHash = await this.dbManager.numberToHash(blockNumber + BIGINT_1)
+        //   if (childHeaderHash !== undefined) {
+        //     const childHeader = await this.dbManager
+        //       .getHeaderSafe(childHeaderHash, blockNumber + BIGINT_1)
+        //       .catch((_e) => undefined)
+        //     isHeadChainStillCanonical =
+        //       childHeader !== undefined && equalsBytes(childHeader.parentHash, blockHash)
+        //   } else {
+        //     isHeadChainStillCanonical = false
+        //   }
+        // } else {
+        //   isHeadChainStillCanonical = true
+        // }
 
-        // set total difficulty in the current context scope
-        if (this._headHeaderHash) {
-          currentTd.header = await this.getTotalDifficulty(this._headHeaderHash)
-        }
-        if (this._headBlockHash) {
-          currentTd.block = await this.getTotalDifficulty(this._headBlockHash)
+        let dbOps: DBOp[] = []
+        dbOps = dbOps.concat(DBSetBlockOrHeader(item))
+        dbOps.push(DBSetHashToNumber(blockHash, blockNumber))
+
+        let parentTd = await this.dbManager
+          .getTotalDifficultySafe(header.parentHash, BigInt(blockNumber) - BIGINT_1)
+          .catch((_e) => undefined)
+        const isOptimistic = parentTd === undefined
+        parentTd = parentTd ?? opts?.parentTd
+        const isComplete = parentTd !== undefined
+
+        if (!isOptimistic) {
+          if (this._validateBlocks && !isGenesis && item instanceof Block) {
+            // this calls into `getBlock`, which is why we cannot lock yet
+            await this.validateBlock(block)
+          }
         }
 
-        // calculate the total difficulty of the new block
-        const parentTd = await this.getParentTD(header)
-        if (!block.isGenesis()) {
-          td += parentTd
-        }
+        // 1. if canonical is explicit false then just dump the block
```

Comment:
What does `canonical === false` mean here?  Just that we got a block that we know isn't canonical but we want to hold on to it "for future reference"?